insights/parsers/initscript.py

```diff
@@ -0,0 +1,104 @@
+"""
+InitScript - files ``/etc/rc.d/init.d/*``
+=========================================
+
+InitScript is a parser for the initscripts in ``/etc/rc.d/init.d``.
+
+Because this parser read multiple files, the initscripts are stored as a list
+within the parser and need to be iterated through in order to find specific
+initscripts.
+
+Examples:
+
+    >>> for initscript in shared[InitScript]: # Parser contains list of all initscripts
+    ...     print "Name:", initscript.file_name
+    ...
+    Name: netconsole
+    Name: rhnsd
+
+"""
+
+from . import ParseException
+from .. import Parser, parser
+import re
+
+CHKCONFIG_REGEX = re.compile(r"^#\s+chkconfig:\s+.*$")
+LSB_REGEX = re.compile(r"^#\s+Provides:\s+.*$")
+
+SHEBANG_REGEX = re.compile(r"^#!\s*/.*$")
+STARTSTOPSTATUS_REGEX = re.compile(r"\b(start|stop|status)\b")
+
+
+class EmptyFileException(ParseException):
+    pass
+
+
+class NotInitscriptException(ParseException):
+    pass
+
+
+@parser("initscript")
+class InitScript(Parser):
+    """
+    Parse initscript files. Each item is a dictionary with following fields:
+
+    Attributes:
+        file_name (str): initscript name
+        file_path (str): initscript path without leading '/'
+        file_content (list): initscript content, line by line
+
+    Because some files may not be real initscripts, to determine whether a file
+    in ``etc/rc.d/init.d/`` is an initscript, the parser checks for
+    ``# chkconfig: <values>`` or ``# Provides: <names>`` strings in the script.
+    If that matches, then it assumes it is an initscript.
+
+    Otherwise, it tries to find out if it is by searching for
+
+    * shebang (e.g. ``#!/bin/bash``) on first line
+    * ``start``/``stop``/``status`` tokens in script
+
+    If 3 or more items are found (half the items searched for + 1), called
+    *confidence* in the code (e.g.  shebang + start + stop), then we assume it
+    is an initscript.
+
+    Otherwise the parser raises a ``ParseException``.
+    """
+
+    def parse_content(self, content):
+        """
+        Raises:
+            EmptyFileException: Raised if file is empty.
+            NotInitscriptException: Raised if likely not an initscript.
+        """
+        self.file_content = content
+
+        # If we find 'chkconfig: XYZ' or 'Provides: NAME', assume we have an
+        # initscript
+        for line in content:
+            if CHKCONFIG_REGEX.match(line) or LSB_REGEX.match(line):
+                return
+
+        # Otherwise, check for hints (shebang + presence of start/stop/status
+        # keywords)
+        #
+        # Allocate 1 point for each item found. And declare as valid if
+        # confidence >= 3 (half of the number of items + 1).
+        confidence = 0
+        try:
+            line = content[0]
+            if SHEBANG_REGEX.match(line):
```

Comment:
I love regexes as much as anyone but you REALLY don't need a regex here.  You can do:
```python
if line.startswith('#!/'):
```

Reply:
The parser will be used with a (for now) DiagInsights rule that will search for calls for "su" which have to be avoided on RHEL7 and later (i.e. with systemd).
Regarding the shebang, "#!/" is not enough, because some people put a space between ! and /.

Reply:
Sure.  Search for '#!'.  Works for me.

Reply:
Still, different because shebang has to be an absolute path, so executable path should start with "/". My idea was to provide something clean, still I won't argue anymore, I have other higher priority tasks, including customer satisfaction.

Regarding your "grep" proposal, I want to parse sosreports from customers.
If there is a way to use system commands on sosreport files, then please explain how to do that.
I can then rewrite most of the current DiagInsights rules into basic greps, which will be faster than using Insights code.